runtime/ops/net.rs

```diff
@@ -528,3 +546,302 @@ fn op_listen(
+  A(String),
+  AAAA(String),
+  ANAME(String),
+  CNAME(String),
+  MX {
+    preference: u16,
+    exchange: String,
+  },
+  PTR(String),
+  SRV {
+    priority: u16,
+    weight: u16,
+    port: u16,
+    target: String,
+  },
+  TXT(Vec<String>),
+}
+
+// TODO(magurotuna): `RecordType` defined in trust_dns_proto doesn't implement Deserialize
+// trait right now, but it's coming in the next version.
+// See https://github.com/bluejekyll/trust-dns/pull/1319
+#[derive(Deserialize, Clone, Copy)]
+enum SupportedRecordType {
+  A,
+  AAAA,
+  ANAME,
+  CNAME,
+  MX,
+  PTR,
+  SRV,
+  TXT,
+}
+
+async fn op_dns_resolve(
+  state: Rc<RefCell<OpState>>,
+  args: Value,
+  _zero_copy: BufVec,
+) -> Result<Value, AnyError> {
+  fn default_port() -> u16 {
+    53
+  }
+
+  #[derive(Deserialize)]
+  #[serde(rename_all = "camelCase")]
+  struct ResolveAddrArgs {
+    query: String,
+    record_type: SupportedRecordType,
+    options: Option<ResolveDnsOption>,
+  }
+
+  #[derive(Deserialize)]
+  #[serde(rename_all = "camelCase")]
+  struct ResolveDnsOption {
+    name_server: Option<NameServer>,
+  }
+
+  impl Into<RecordType> for SupportedRecordType {
+    fn into(self) -> RecordType {
+      use SupportedRecordType::*;
+      match self {
+        A => RecordType::A,
+        AAAA => RecordType::AAAA,
+        ANAME => RecordType::ANAME,
+        CNAME => RecordType::CNAME,
+        MX => RecordType::MX,
+        PTR => RecordType::PTR,
+        SRV => RecordType::SRV,
+        TXT => RecordType::TXT,
+      }
+    }
+  }
+
+  #[derive(Deserialize, Clone, Copy)]
+  enum Protocol {
+    UDP,
+    TCP,
+  }
+
+  impl Default for Protocol {
+    fn default() -> Self {
+      Self::UDP
+    }
+  }
+
+  impl Into<TrustProtocol> for Protocol {
+    fn into(self) -> TrustProtocol {
+      use Protocol::*;
+      match self {
+        UDP => TrustProtocol::Udp,
+        TCP => TrustProtocol::Tcp,
+      }
+    }
+  }
+
+  #[derive(Deserialize)]
+  #[serde(rename_all = "camelCase")]
+  struct NameServer {
+    ip_addr: String,
+    #[serde(default = "default_port")]
+    port: u16,
+    #[serde(default)]
+    protocol: Protocol,
+  }
+
+  let ResolveAddrArgs {
+    query,
+    record_type,
+    options,
+  } = serde_json::from_value(args)?;
+
+  // Check for permission. If name server option is provided, use them for checking.
+  // Otherwise, checks if `allow-net` is present without any specific allowlist.
+  {
+    let perm_args = options
+      .as_ref()
+      .and_then(|opt| opt.name_server.as_ref())
+      .map(|ns| (&ns.ip_addr, ns.port));
+
+    let s = state.borrow();
+    let perm = s.borrow::<Permissions>();
+
+    if let Some((ip_addr, port)) = perm_args {
+      perm.check_net(ip_addr, port)?;
+    } else {
+      perm.check_net_all()?;
+    }
+  }
+
+  let resolver = if let Some(options) = options {
+    let conf = if let Some(name_server) = options.name_server {
+      let mut c = ResolverConfig::new();
+      c.add_name_server(NameServerConfig {
```

Comment:
This adds the server to the end of the list but I think that means the default ones will be queried first. It'd be better to construct the Resolverconfig like this:
```rust
// Don't try other servers on SERVFAIL or NXDOMAIN - I *think* that's appropriate
let trust_nx_responses = true;
let group = NameServerConfigGroup::from_ips_clear(&[name_server], 53, trust_nx_responses);
let c = Resolverconfig::from_parts(None, vec![], group);
```

Reply:
> This adds the server to the end of the list

Oh I didn't know that, I will adopt your way and see if it works as we expect.